From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lei Zhang <thestig@chromium.org>
Date: Fri, 2 Oct 2026 11:48:28 -0700
Subject: [M156] [PDF] Avoid using invalid link handle in PDFiumPage

Original change's description:
> [PDF] Avoid using invalid link handle in PDFiumPage
>
> Page loading can trigger JS that invalidates a link handle that
> PDFiumPage::PopulateAnnotationLinks() uses. Avoid this potential problem
> by first using the link handle to retrieve all the necessary
> information, before accessing the page that may invalidate the handle.
> Add comments to warn about the potential handle lifetime issues.
>
> Bug: 567910530
> Change-Id: Ic797764cf397a1e3986617faadf358eba1d32472
> Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8492452
> Commit-Queue: Lei Zhang <thestig@chromium.org>
> Reviewed-by: Tom Sepez <tsepez@chromium.org>
> Cr-Commit-Position: refs/heads/main@{#1709135}

(cherry picked from commit ff834f798fe3b7649892f36a6a9cb14ad6f01672)

Bug: 568628162,567910530
Change-Id: Ic797764cf397a1e3986617faadf358eba1d32472
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8503366
Commit-Queue: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
Bot-Commit: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
Auto-Submit: chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com <chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com>
Cr-Commit-Position: refs/branch-heads/8078@{#227}
Cr-Branched-From: c50edc4214d81e4f944c3415d55104a49e46d214-refs/heads/main@{#1706539}

diff --git a/pdf/pdfium/pdfium_page.cc b/pdf/pdfium/pdfium_page.cc
index 30ecfc42e725f6823c7efdf42edc336ae1a7f509..ef209e39daf16de43a029408c98831f5d73dee57 100644
--- a/pdf/pdfium/pdfium_page.cc
+++ b/pdf/pdfium/pdfium_page.cc
@@ -1584,11 +1584,6 @@ void PDFiumPage::PopulateAnnotationLinks() {
   // deletions below prevents one of them from erasing another.
   std::vector<Link> annotation_links;
   while (FPDFLink_Enumerate(page, &start_pos, &link_annot)) {
-    Link link;
-    Area area = GetLinkTarget(link_annot, &link.target);
-    if (area == NONSELECTABLE_AREA)
-      continue;
-
     PdfRect link_rect;
     if (!FPDFLink_GetAnnotRect(link_annot, &FsRectFFromPdfRect(link_rect))) {
       continue;
@@ -1598,6 +1593,7 @@ void PDFiumPage::PopulateAnnotationLinks() {
     // flipped. Swap the coordinates before further processing.
     link_rect.Normalize();
 
+    Link link;
     int quad_point_count = FPDFLink_CountQuadPoints(link_annot);
     // Calculate the bounds of link using the quad points data.
     // If quad points for link is not present then use
@@ -1618,6 +1614,15 @@ void PDFiumPage::PopulateAnnotationLinks() {
                                                  PageOrientation::kOriginal));
     }
 
+    // WARNING: Do not use `link_annot` after this call. It may have been
+    // invalidated.
+    Area area = GetLinkTarget(link_annot, &link.target);
+    if (area == NONSELECTABLE_AREA) {
+      // It is unfortunate that all the work above may get thrown away due to
+      // how `link_annot` access have to be arranged to be safe.
+      continue;
+    }
+
     // Calculate underlying text range of link.
     GetUnderlyingTextRangeForRect(
         gfx::RectF(link_rect.left(), link_rect.bottom(), link_rect.width(),
